### PR TITLE
Initial implementations of UrlResolvers and UrlLoaders

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

**Comment by justinfagnani** [comment]
cc @garlicnation @azakus @FredKSchott

File: package.json [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  },
+  "dependencies": {},
+  "devDependencies": {
+    "chai": "^3.5.0",
+    "fs-extra": "^0.30.0",
+    "gulp": "^3.9.1",
+    "gulp-rename": "^1.2.2",
+    "gulp-typescript": "^2.13.4",
+    "gulp-typings": "^2.0.0",
+    "run-sequence": "^1.2.0"
```
Comment (dfreedm) [inline package.json]: gulp-mocha is missing

File: package.json [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  },
+  "scripts": {
+    "test": "gulp test"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/Polymer/file-loader.git"
+  },
+  "author": "",
+  "license": "ISC",
```
Comment (dfreedm) [inline package.json]: BSD-3-Clause

**Comment by justinfagnani** [comment]
FYI, the plan as of now is to also add an AppUrlResolver that doesn't support `../` URLs, so that we don't accidentally analyze files for app-style projects. At some point in the tool chain we'll need to differentiate between app and element and thread the relevant settings down.

File: package.json (hunk 1 above)
Comment (justinfagnani) [inline package.json]: done

File: package.json (hunk 2 above)
Comment (justinfagnani) [inline package.json]: done

File: src/url-loader.ts [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+  /**
+   * Returns `true` if this loader can load the given `url`.
+   */
+  canLoad(url: string): boolean;
+
+  /**
+   * Reads a file from `url`.
+   *
+   * This should only be called if `handlesUrl` returns `true` for `url`.
```
Comment (garlicnation) [inline src/url-loader.ts]: change to `canLoad`?

File: src/package-url-resolver.ts [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ * @license
+ * Copyright (c) 2016 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at http://polymer.github.io/PATENTS.txt
+ */
+
+import {posix as pathlib} from 'path';
```
Comment (garlicnation) [inline src/package-url-resolver.ts:11]: Will this work on windows?

File: src/package-url-resolver.ts [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+    let urlObject = parseUrl(url);
+    let pathname = pathlib.normalize(decodeURIComponent(urlObject.pathname));
+
+    if (!this._isValid(urlObject, pathname)) {
+      throw new Error(`Invalid URL ${url}`);
+    }
+
+    // If the path points to a sibling directory, resolve it to the
+    // component directory
+    if (pathname.startsWith('../')) {
```
Comment (garlicnation) [inline src/package-url-resolver.ts:55]: The path is normalized - how could it start with ../?

I suspect you'll have to do the `isSiblingOrAunt` algorithm with the path parts.

File: test/package-url-resolver_test.js [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    test('canResolve is false for a cousin URL', () => {
+      assert.isFalse(new PackageUrlResolver().canResolve('../../foo/foo.html'));
+    });
+
+    test('canResolve is false for URL with a hostname', () => {
+      assert.isFalse(new PackageUrlResolver().canResolve('http://abc.xyz/foo.html'));
+    });
+
+    test('canResolve is true for a URL with the right hostname', () => {
+      assert.isTrue(new PackageUrlResolver({
```
Comment (garlicnation) [inline test/package-url-resolver_test.js]: more cases:
 `canResolve('foo.html')`
`canResolve('/foo.html')`
`canResolve('../foo.html')`

File: src/package-url-resolver.ts (hunk 4 above)
Comment (justinfagnani) [inline src/package-url-resolver.ts:11]: Yes, the UrlResolver is only mapping URL->URL and so shouldn't use Windows separators. UrlLoaders should use the default path library though.

File: src/package-url-resolver.ts (hunk 5 above)
Comment (justinfagnani) [inline src/package-url-resolver.ts:55]: normalization doesn't remove `../` segments. It collapses consecutive `/`s and `./`s

File: src/url-loader.ts (hunk 3 above)
Comment (justinfagnani) [inline src/url-loader.ts]: done

File: test/package-url-resolver_test.js (hunk 6 above)
Comment (justinfagnani) [inline test/package-url-resolver_test.js]: done

**Comment by justinfagnani** [comment]
@garlicnation PTAL

(and for some reason, I can't assign you to the PR) edit: oops, didn't add a team...

File: src/package-url-resolver.ts (hunk 5 above)
Comment (garlicnation) [inline src/package-url-resolver.ts:55]: What about absolute URLS?

File: src/url-loader.ts (hunk 3 above)
Comment (garlicnation) [inline src/url-loader.ts]: Doesn't show up in my diff

File: src/package-url-resolver.ts (hunk 5 above)
Comment (justinfagnani) [inline src/package-url-resolver.ts:55]: normalization preserves the leading `/`

File: src/url-loader.ts (hunk 3 above)
Comment (justinfagnani) [inline src/url-loader.ts]: oops, pushed

File: test/package-url-resolver_test.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+    test('throws for a URL with a hostname', () => {
+      assert.throws(() => new PackageUrlResolver().resolve('http://abc.xyz/foo.html'));
+    });
+
+    test('resolves a URL with the right hostname', () => {
+      let resolver = new PackageUrlResolver({
+        hostname: 'abc.xyz',
+      });
+      assert.equal('foo.html', resolver.resolve('http://abc.xyz/foo.html'));
```
Comment (garlicnation) [inline test/package-url-resolver_test.js]: resolve always must be called with a hostname? html imports will have relative URLs - where do you expect a user to attach the hostname?

Comment (justinfagnani) [inline test/package-url-resolver_test.js]: No, see the first test: 'resolves an in-package URL'. This test is only for when you explicitly give the resolver a hostname to handle.

Comment (garlicnation) [inline test/package-url-resolver_test.js]: I would like to see a test like:

```
let resolver = new PackageUrlResolver({
         hostname: 'abc.xyz',
});
assert.equal('someString', resolver.resolve('foo.html'))
```

Comment (justinfagnani) [inline test/package-url-resolver_test.js]: done

File: src/package-url-resolver.ts (hunk 5 above)
Comment (garlicnation) [inline src/package-url-resolver.ts:55]: Exactly - /components/dependency/dependency-import.html should resolve to bower_components if resolving with a base of /components/element/element.html

Comment (justinfagnani) [inline src/package-url-resolver.ts:55]: I don't follow...

Comment (garlicnation) [inline src/package-url-resolver.ts:55]: This is the case I had in mind, roughly: https://gist.github.com/garlicnation/edc7ebba8eb89b1b5d93a1fdab229f73

Comment (justinfagnani) [inline src/package-url-resolver.ts:55]: That would be a different resolver implementation that has an opinion that `/components` should mean `/bower_components`. This resolver is very simple and only has the opinion that sibling directories of the (implied) root map to the `componentDir`

**Pull request merged**